src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,454 @@
+#
+
+import json
+from enum import Enum
+from typing import List, Optional, Tuple
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.error_message import SnowparkClientExceptionMessages
+from snowflake.snowpark.types import (
+    IntegerType,
+    StringType,
+    StructField,
+    StructType,
+    VariantType,
+)
+
+
+class LineageDirection(Enum):
+    """
+    Directions for tracing the lineage.
+
+    Attributes:
+        DOWNSTREAM (str): Represents the downstream direction in lineage tracing.
+        UPSTREAM (str): Represents the upstream direction in lineage tracing.
+        BOTH (str): Represents both upstream and downstream direction in lineage tracing.
+    """
+
+    DOWNSTREAM = "downstream"
+    UPSTREAM = "upstream"
+    BOTH = "both"
+
+    def __str__(self):
+        return self.value
+
+
+class _EdgeType(Enum):
+    """
+    Types of edges for lineage tracing.
+    """
+
+    DATA_LINEAGE = "DATA_LINEAGE"
+    OBJECT_DEPENDENCY = "OBJECT_DEPENDENCY"
+
+    @classmethod
+    def list(cls):
+        return [member.value for member in cls]
+
+
+class _ObjectField:
+    """
+    Defines static fields used to reference object properties in DGQL query and response.
+    """
+
+    DOMAIN = "domain"
+    REFINED_DOMAIN = "refinedDomain"
+    USER_DOMAIN = "userDomain"
+    NAME = "name"
+    PROPERTIES = "properties"
+    SCHEMA = "schema"
+    DB = "db"
+    STATUS = "status"
+    CREATED_ON = "createdOn"
+    PARENT_NAME = "ParentName"
+    VERSION = "version"
+
+    # A list of fileds queried on each object in the lineage.
+    GRAPH_ENTITY_PROPERTIES = [
+        DOMAIN,
+        REFINED_DOMAIN,
+        USER_DOMAIN,
+        NAME,
+        PROPERTIES,
+        SCHEMA,
+        DB,
+        STATUS,
+        CREATED_ON,
+    ]
+
+
+class _DGQLFields:
+    """
+    Contains static definitions of field names used in DGQL queries and responses.
+    """
+
+    DATA = "data"
+    NODE = "V"
+    EDGE = "E"
+    SOURCE = "S"
+    TARGET = "T"
+    OUT = "OUT"
+    IN = "IN"
+
+
+class _UserDomain:
+    """
+    Domains in the user context that logically maps to different snowflake objects.
+    """
+
+    FEATURE_VIEW = "FEATURE_VIEW"
+    MODEL = "MODULE"
+
+
+class _SnowflakeDomain:
+    """
+    Snowflake object domains relevant for querying lineage.
+    Note: This is a subset and does not include all possible domains.
+    """
+
+    TABLE = "TABLE"
+    MODULE = "MODULE"
+    DATASET = "DATASET"
+
+
+class Lineage:
+    """
+    Provides methods for exploring lineage of Snowflake objects.
+    To access an object of this class, use :attr:`Session.lineage`.
+    """
+
+    def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
+        self._session = session
+        self.user_to_system_domain_map = {
+            _UserDomain.FEATURE_VIEW: _SnowflakeDomain.TABLE,
+            _UserDomain.MODEL: _SnowflakeDomain.MODULE,
+        }
+        self.versioned_object_domains = {
```

Comment:
Are you intending to expose `user_to_system_domain_map` and `versioned_object_domains` to users?

Reply:
Nope. made it _*